### DISCUSSION TITLE
slack_notify script in shared-lib not working for some pipelines

### DISCUSSION THREAD

**Discussion by vj4a** [Sunbird-Ed/Community#652] [Installation] [answered]
Hello @santhosh-tg , @gandham-santhosh ,

In some pipelines like **ArtifactUpload/Core/Assessment:** , slack_notify stage is passing as expected irrespective of whether required environment variables is passed or not but in some  like **ArtifactUpload/Core/AdminUtils**, it gives the below exception,

**Org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 8a11b382-45fd-4cdc-a70e-2e9c05661929
hudson.remoting.ProxyException: org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object '' with class 'java.lang.String' to class 'int'**

We have tried the shared-lib present in the repo [dicdiksha/sunbird-devops](https://github.com/dicdiksha/sunbird-devops/blob/shared-lib/vars/slack_notify.groovy) where it looks like it has been fixed with this [commit](https://github.com/dicdiksha/sunbird-devops/commit/b0898582d00ba72538a8469e69632ad82b63715c) but we are still facing the same error.

We'll look into the same. Kindly let us know if there is any fix already available.

Thanks

**Comment by santhosh-tg** [comment]
Hi @vj4a We haven't faced this issue before. Even with recent Installation at OCI we didn't have this issue.

It might be related to the Jenkins upgrade / plugins version but not sure though. You may want debug further to confirm.

https://github.com/heungheung https://github.com/ddevadat Please add your inputs if any.

**Reply by vj4a** [reply]
Okay @santhosh-tg , Thanks for the info. We'll check it further.
We have started to work on the service provision pipelines in parallel as the artifact upload steps have all passed.
Are we right that that there is no dependency on these notification steps in the upcoming stages.

**Reply by santhosh-tg** [reply]
yes, no dependency. Its just about sending slack notifications about the job status.

**Reply by heungheung** [reply]
I have seen this before - the error is self explained - the problem is when you crafting the slack message or presetting the values, you try to put a string object as an integer. At that time, I was quite surprised no one asked similar question . Anyway, this depends how you actually did the slack integration, the plug-in version, groovy and jenkins version - they are mixed together

My quick workaround at that time is create my own shared lib and changed the slack notification function - again, this might depends on the jenkins version and slack plugin version - the commit you shared is the change I've made at that time - the jenkins version is different, that might not work in your case. 

Sorry, I don't have a quick solution - even at that time, I was trying different combination - I might have another private repo got different changes but it was long time ago, need more time to check if I can found it out....

**Reply by gandham-santhosh** [reply]
@vj4a have you fixed this issue in your end? If so please share the details

**Reply by vj4a** [reply]
Not yet @gandham-santhosh , we are checking on it.

Except for Artifact Uploads/Core/AdminUtils, all the other notify errors are pertinent to uploads to the storage bucket and not container registry which is a pattern we have observed.  

To rule out whether this is due to Jenkins latest version, we'll test that too.
We'll update here once we have a fix for it.

**Reply by heungheung** [reply]
I don't know whether it will work - but just my guess

when enter the function
```
def call(String buildStatus, String release_tag=null, String jobName=null, int buildNumber=0, String jobUrl=null)
```
buildNumber is an integer, with 0 as default if whoever calling it does not provide the value

in this line 
```
 if (buildNumber == 0)
```
we are doing integer comparison. In fact, if **true**
```
buildNumber = env.BUILD_NUMBER
```
while all env.* are actually string - error happens when no values is pass to the function OR the passed value is a 0

things you can try - (it is groovy, not sure how that works in jenkins)  do a def or a int, in the very begining - either
```
def myBuildNumber = env.BUILD_NUMBER.toInteger()
```
or
```
int myBuildNumber = env.BUILD_NUMBER
```
this is a bit different because the cast happened even before an if then, it is in the declaration/definition - again, those code are groovy - I haven't tried inside jenkins. If this dec/def works, you should be able to do 
```
                if (buildNumber == 0)
                    buildNumber = myBuildNumber 
```
where myBuildNumber is an integer - so that integer to integer assignment should NOT throw error

**Reply by heungheung** [reply]
or maybe just
```
                if (buildNumber == 0)
                    buildNumber = env.BUILD_NUMBER.toInteger()
```
if it is supported in grovvy in jenkins

**Reply by vj4a** [reply, marked as answer]
@heungheung , Thanks for the info, the typecasting was the issue. Also, in artifact upload to storage buckets, the build number is null which is to be checked further. For now, this is a workaround for the exception.
```
                if (buildNumber == 0)
                    env.BUILD_NUMBER = env.BUILD_NUMBER == "" ? "0" : env.BUILD_NUMBER
                    buildNumber = env.BUILD_NUMBER.toInteger()
```

**Comment by mohitga** [comment]
@vj4a  is this issue resolved?

**Reply by vj4a** [reply]
@mohitga , This is resolved. Marking as closed.
